### DISCUSSION TITLE
Porting RPC commands to LSP

### DISCUSSION THREAD

**Discussion by bart-jaskulski** [phpactor/phpactor#2782] [Ideas]
Do you believe it would be beneficial and also simply possible to port existing RPC commands to LSP protocol as code actions?

As I view it, this might contribute to easier access to particular actions without need for editor plugins like phpactor.nvim. Although I am not sure how it would affect the readability of code actions menu (there would be much more actions too choose from at request), and how does amount of code action providers affect the performance of phpactor (I personaly notice a delay when initiating a code actions menu from nvim).

**Comment by dantleech** [comment]
which actions do you have in mind? all the code transformations are in the LSP. class moving is _not_ (mainly because it works in a whole different way) and maybe one or two others but there is practically only one or two concrete reasons ot use the "RPC" system, it's 95% obsolete.

**Reply by bart-jaskulski** [reply]
Hmm... I might blindly assumed, that all actions listed in [phpactor.nvim](https://github.com/gbprod/phpactor.nvim?tab=readme-ov-file#-features) are missing LSP implementation. Namely:

```
PhpActor [
  class_inflect
  context_menu
  expand_class
  generate_accessor
  change_visibility
  copy_class
  import_class
  import_missing_classes
  move_class
  navigate
  new_class
  transform
  update
  config
  status
  cache_clear
  copy_fcqn
  lsp/status
  lsp/reindex
  lsp/debug/config
  lsp/debug/config
  blackfire/start
  blackfire/stop
]
```

But it seems it is only true for `change_visibility`, `generate_accessor` and `navigate` maybe.

**Reply by dantleech** [reply]
yeah, some of those _are_ LSP actions.

fwiw the ones that have no direct counterpart:

- `class_inflect`: i think this generates interfaces from classes
- `context_menu`: replaced by LSP
- `change_visibility`: technically still on the list to port to LSP, although tbh I never bothered as it's so easy to do.
- `{move,copy}_class`: There is some support for moving classes in LSP, although the "RPC" version does this _on disk_ and I prefer it tbh.
- `navigate`: not really something that fits into vanilla LSP - could be integrated as a custom LSP action and have support provided in the client (e.g. phpactor-vscode)
